File: mqtt/clients/mqtt-test-client.js
#!/usr/bin/env node
/**
 * ESP32 MQTT Test Client (Node.js)
 * Interactive CLI tool for testing MQTT communication with ESP32
 */

const mqtt = require('mqtt');
const readline = require('readline');

// MQTT Configuration
const BROKER_URL = 'mqtt://broker.hivemq.com:1883';

// MQTT Topics
const TOPICS = {
    LED_CONTROL: 'esp32/led/control',
    LED_STATUS: 'esp32/led/status',
    DEVICE_STATUS: 'esp32/device/status',
    DEVICE_COMMAND: 'esp32/device/command'
};

class ESP32MqttClient {
    constructor() {
        this.client = null;
        this.connected = false;
        this.setupReadline();
    }

    setupReadline() {
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    connect() {
        console.log(`Connecting to MQTT broker: ${BROKER_URL}`);
        
        this.client = mqtt.connect(BROKER_URL);

        this.client.on('connect', () => {
            console.log('Connected to MQTT broker');
            console.log('Subscribing to ESP32 topics...');
            
            // Subscribe to response topics
            this.client.subscribe([
                TOPICS.LED_STATUS,
                TOPICS.DEVICE_STATUS
            ], (err) => {
                if (!err) {
                    console.log(`   - ${TOPICS.LED_STATUS}`);
                    console.log(`   - ${TOPICS.DEVICE_STATUS}`);
                    this.connected = true;
                    this.showHelp();
                    this.startCommandLoop();
                } else {
                    console.log('Subscription failed:', err.message);
                }
            });
        });

        this.client.on('message', (topic, message) => {
            this.handleMessage(topic, message);
        });

        this.client.on('error', (error) => {
            console.log('MQTT Error:', error.message);
        });

        this.client.on('close', () => {
            this.connected = false;
            console.log('Disconnected from MQTT broker');
        });
    }

    handleMessage(topic, message) {
        const timestamp = new Date().toLocaleTimeString();
        
        try {
            const data = JSON.parse(message.toString());
            
            if (topic === TOPICS.LED_STATUS) {
                const state = data.led_state;
                const msg = data.message || '';
                console.log(`[${timestamp}] LED Status: ${state} (${msg})`);
                
            } else if (topic === TOPICS.DEVICE_STATUS) {
                console.log(`[${timestamp}] Device Status:`);
                console.log(`   Device ID: ${data.device_id || 'unknown'}`);
                console.log(`   IP Address: ${data.ip || 'unknown'}`);
                console.log(`   Uptime: ${data.uptime || 0}s`);
                console.log(`   Free Heap: ${data.heap || 0} bytes`);
                console.log(`   LED State: ${data.led_state}`);
                console.log(`   MQTT Connected: ${data.mqtt_connected}`);
            }
        } catch (error) {
            // Handle plain text messages
            console.log(`[${timestamp}] ${topic}: ${message.toString()}`);
        }
        
        this.showPrompt();
    }

    controlLED(state) {
        if (!this.connected) {
            console.log('Not connected to MQTT broker');
            return;
        }

        const payload = JSON.stringify({ state: state });
        this.client.publish(TOPICS.LED_CONTROL, payload, (err) => {
            if (!err) {
                const action = state ? 'ON' : 'OFF';
                console.log(`Sent LED command: ${action}`);
            } else {
                console.log('Failed to send LED command:', err.message);
            }
            this.showPrompt();
        });
    }

    requestStatus() {
        if (!this.connected) {
            console.log('Not connected to MQTT broker');
            return;
        }

        this.client.publish(TOPICS.DEVICE_COMMAND, 'status', (err) => {
            if (!err) {
                console.log('Requested device status');
            } else {
                console.log('Failed to request status:', err.message);
            }
            this.showPrompt();
        });
    }

    restartDevice() {
        if (!this.connected) {
            console.log('Not connected to MQTT broker');
            return;
        }

        this.rl.question('Are you sure you want to restart the ESP32? (y/n): ', (answer) => {
            if (answer.toLowerCase() === 'y') {
                this.client.publish(TOPICS.DEVICE_COMMAND, 'restart', (err) => {
                    if (!err) {
                        console.log('Sent restart command');
                    } else {
                        console.log('Failed to send restart command:', err.message);
                    }
                    this.showPrompt();
                });
            } else {
                console.log('Restart cancelled');
                this.showPrompt();
            }
        });
    }

    showHelp() {
        console.log('\nAvailable Commands:');
        console.log('   on      - Turn LED ON');
        console.log('   off     - Turn LED OFF');
        console.log('   status  - Request device status');
        console.log('   restart - Restart ESP32 device');
        console.log('   help    - Show this help');
        console.log('   quit    - Exit program');
        console.log();
    }

    showPrompt() {
        process.stdout.write('Enter command: ');
    }

    startCommandLoop() {
        this.showPrompt();
        
        this.rl.on('line', (input) => {
            const command = input.trim().toLowerCase();
            
            switch (command) {
                case 'on':
                    this.controlLED(true);
                    break;
                    
                case 'off':
                    this.controlLED(false);
                    break;
                    
                case 'status':
                    this.requestStatus();
                    break;
                    
                case 'restart':
                    this.restartDevice();
                    break;
                    
                case 'help':
                    this.showHelp();
                    this.showPrompt();
                    break;
                    
                case 'quit':
                case 'exit':
                case 'q':
                    this.disconnect();
                    break;
                    
                default:
                    if (command) {
                        console.log(`Unknown command: '${command}'. Type 'help' for available commands.`);
                    }
                    this.showPrompt();
                    break;
            }
        });
    }

    disconnect() {
        console.log('Disconnecting...');
        if (this.client) {
            this.client.end();
        }
        this.rl.close();
        console.log('Goodbye!');
        process.exit(0);
    }
}

// Check if mqtt module is available
try {
    require('mqtt');
} catch (error) {
    console.log('Error: mqtt module not found');
    console.log('Install it using: npm install mqtt');
    process.exit(1);
}

// Handle process interruption
process.on('SIGINT', () => {
    console.log('\nInterrupted by user');
    process.exit(0);
});

// Start the application
console.log('ESP32 MQTT Test Client (Node.js)');
console.log('=' .repeat(40));

const client = new ESP32MqttClient();
client.connect();

File: mqtt/clients/mqtt-test-client.py
#!/usr/bin/env python3
"""
ESP32 MQTT Test Client
Simple Python script to test MQTT communication with ESP32 device
"""

import paho.mqtt.client as mqtt
import json
import time
import threading
from datetime import datetime

# MQTT Configuration
BROKER = "broker.hivemq.com"
PORT = 1883
KEEPALIVE = 60

# MQTT Topics
TOPIC_LED_CONTROL = "esp32/led/control"
TOPIC_LED_STATUS = "esp32/led/status"
TOPIC_DEVICE_STATUS = "esp32/device/status"
TOPIC_DEVICE_COMMAND = "esp32/device/command"

class ESP32MqttClient:
    def __init__(self):
        self.client = mqtt.Client()
        self.client.on_connect = self.on_connect
        self.client.on_message = self.on_message
        self.client.on_disconnect = self.on_disconnect
        self.connected = False
        
    def on_connect(self, client, userdata, flags, rc):
        if rc == 0:
            print(f"Connected to MQTT broker: {BROKER}:{PORT}")
            print("Subscribing to ESP32 topics...")
            
            # Subscribe to ESP32 response topics
            client.subscribe(TOPIC_LED_STATUS)
            client.subscribe(TOPIC_DEVICE_STATUS)
            
            print(f"   - {TOPIC_LED_STATUS}")
            print(f"   - {TOPIC_DEVICE_STATUS}")
            self.connected = True
        else:
            print(f"Failed to connect, return code {rc}")
            
    def on_message(self, client, userdata, msg):
        topic = msg.topic
        try:
            # Try to parse as JSON
            message = json.loads(msg.payload.decode())
            timestamp = datetime.now().strftime("%H:%M:%S")
            
            if topic == TOPIC_LED_STATUS:
                state = message.get('led_state', 'unknown')
                msg_text = message.get('message', '')
                print(f"[{timestamp}] LED Status: {state} ({msg_text})")
                
            elif topic == TOPIC_DEVICE_STATUS:
                device_id = message.get('device_id', 'unknown')
                ip = message.get('ip', 'unknown')
                uptime = message.get('uptime', 0)
                heap = message.get('heap', 0)
                led_state = message.get('led_state', False)
                mqtt_connected = message.get('mqtt_connected', False)
                
                print(f"[{timestamp}] Device Status:")
                print(f"   Device ID: {device_id}")
                print(f"   IP Address: {ip}")
                print(f"   Uptime: {uptime}s")
                print(f"   Free Heap: {heap} bytes")
                print(f"   LED State: {led_state}")
                print(f"   MQTT Connected: {mqtt_connected}")
                
        except json.JSONDecodeError:
            # Handle plain text messages
            message = msg.payload.decode()
            timestamp = datetime.now().strftime("%H:%M:%S")
            print(f"[{timestamp}] {topic}: {message}")
            
    def on_disconnect(self, client, userdata, rc):
        self.connected = False
        if rc != 0:
            print("Unexpected MQTT disconnection. Attempting to reconnect...")
        else:
            print("Disconnected from MQTT broker")
            
    def connect(self):
        try:
            print(f"Connecting to MQTT broker {BROKER}:{PORT}...")
            self.client.connect(BROKER, PORT, KEEPALIVE)
            self.client.loop_start()
        except Exception as e:
            print(f"Connection error: {e}")
            
    def disconnect(self):
        self.client.loop_stop()
        self.client.disconnect()
        
    def control_led(self, state):
        if not self.connected:
            print("Not connected to MQTT broker")
            return False
            
        payload = json.dumps({"state": state})
        result = self.client.publish(TOPIC_LED_CONTROL, payload)
        
        if result.rc == mqtt.MQTT_ERR_SUCCESS:
            action = "ON" if state else "OFF"
            print(f"Sent LED command: {action}")
            return True
        else:
            print(f"Failed to send LED command")
            return False
            
    def request_status(self):
        if not self.connected:
            print("Not connected to MQTT broker")
            return False
            
        result = self.client.publish(TOPIC_DEVICE_COMMAND, "status")
        
        if result.rc == mqtt.MQTT_ERR_SUCCESS:
            print("Requested device status")
            return True
        else:
            print("Failed to request status")
            return False
            
    def restart_device(self):
        if not self.connected:
            print("Not connected to MQTT broker")
            return False
            
        result = self.client.publish(TOPIC_DEVICE_COMMAND, "restart")
        
        if result.rc == mqtt.MQTT_ERR_SUCCESS:
            print("Sent restart command")
            return True
        else:
            print("Failed to send restart command")
            return False

def print_help():
    print("\nAvailable Commands:")
    print("   on      - Turn LED ON")
    print("   off     - Turn LED OFF")
    print("   status  - Request device status")
    print("   restart - Restart ESP32 device")
    print("   help    - Show this help")
    print("   quit    - Exit program")
    print()

def main():
    print("ESP32 MQTT Test Client")
    print("=" * 40)
    
    # Create and connect client
    client = ESP32MqttClient()
    client.connect()
    
    # Wait for connection
    time.sleep(2)
    
    if not client.connected:
        print("Failed to connect. Please check:")
        print("   - Internet connection")
        print("   - MQTT broker availability")
        print("   - Firewall settings")
        return
        
    print_help()
    
    try:
        while True:
            command = input("Enter command: ").lower().strip()
            
            if command == "on":
                client.control_led(True)
                
            elif command == "off":
                client.control_led(False)
                
            elif command == "status":
                client.request_status()
                
            elif command == "restart":
                confirm = input("Are you sure you want to restart the ESP32? (y/n): ")
                if confirm.lower() == 'y':
                    client.restart_device()
                else:
                    print("Restart cancelled")
                    
            elif command == "help":
                print_help()
                
            elif command in ["quit", "exit", "q"]:
                break
                
            else:
                print(f"Unknown command: '{command}'. Type 'help' for available commands.")
                
    except KeyboardInterrupt:
        print("\nInterrupted by user")
        
    finally:
        print("Disconnecting...")
        client.disconnect()
        print("Goodbye!")

if __name__ == "__main__":
    # Check if paho-mqtt is installed
    try:
        import paho.mqtt.client as mqtt
    except ImportError:
        print("Error: paho-mqtt library not found")
        print("Install it using: pip install paho-mqtt")
        exit(1)
        
    main()